### PR TITLE
feat(windows): Rework of profile installation - Engine

### PR DESCRIPTION
**Pull request by mcdurdin**
This PR replaces #3483.

Two follow-up PRs implement the Keyman for Windows changes and a support tool.

Re-implements the Keyman Engine interfaces for registration and installation of TIPs.

With this new model of keyboard installation, the existing `.Install()` methods are deprecated, because the install is now multiple steps, some at user privilege, and some at machine privilege (elevation):

1. Find the target language ID for a given BCP 47 code, including installation, if necessary of a transient language ID and a temporary Windows keyboard to 'hold' that transient language ID.
2. Elevate and register the keyboard TIP for the target language ID, if not already registered.
3. Back in user context, install the now-registered TIP for the current user.

At keyboard install time, Keyman registers the associated TIP for each language provided in the .kmp/.kmx metadata, as well as for the [four transient language IDs](https://github.com/keymanapp/keyman/wiki/Making-sense-of-the-Windows-Layout-Registration#windows-8). This means that elevation is not required for normal language installation scenarios after the initial Keyman keyboard install.

**Note:** Keyman uses the term "Install" to describe adding a keyboard into the system (as you would 'install' an app). This requires elevation. Windows uses the term "Register" for its part of this process, where an input method is added to the system (but not made available for use). "Install" is used by Windows to describe making a registered keyboard available for a user. Sorry for the confusion folks... I'm probably going to end up mixing up the terms periodically.

The core work is in two units:

* Keyman.System.Process.KPInstallKeyboardLanguage.pas
* Keyman.System.Process.KPUninstallKeyboardLanguage.pas

These two units are replacements for kpinstallkeyboardlanguageprofile.pas and kpuninstallkeyboardlanguageprofile.pas. They diverged far enough to make it worth rewriting them completely.

See also #3512 for work outstanding.

### CODE REVIEW

File: windows/src/engine/kmcomapi/com/keyboardlanguages/keymankeyboardlanguageinstalled.pas [hunk 1]
```diff
@@ -34,10 +34,11 @@ interface
   msctf;
 
 type
-  TKeymanKeyboardLanguageInstalled = class(TKeymanKeyboardLanguage, IKeymanKeyboardLanguageInstalled, IIntKeymanKeyboardLanguage)   // I4376
+  TKeymanKeyboardLanguageInstalled = class(TKeymanKeyboardLanguage, IKeymanKeyboardLanguageInstalled, IKeymanKeyboardLanguageInstalled2, IIntKeymanKeyboardLanguage)   // I4376
```
Comment (darcywong00) [inline windows/src/engine/kmcomapi/com/keyboardlanguages/keymankeyboardlanguageinstalled.pas:37]: What's the difference between `IKeymanKeyboardLanguageInstalled` and  `IKeymanKeyboardLanguageInstalled2`?

Comment (mcdurdin) [reply windows/src/engine/kmcomapi/com/keyboardlanguages/keymankeyboardlanguageinstalled.pas:37]: `IKeymanKeyboardLanguageInstalled2` is a newer version of the `IKeymanKeyboardLanguageInstalled` interface (and inherits from it). Per COM, you should not modify an interface once it is published except by following [certain complex rules](https://docs.microsoft.com/en-us/windows/win32/rpc/the-versioning-theory-for-rpc-and-com) which make life more difficult for the client and the library, so instead we introduce a new version of the interface with the additional methods we need, which is easy to understand.

I guess in theory we could rename it to `IKeymanKeyboardLanguageInstalled14` to indicate which version the interface appears in, but sequential numbering is common practice.

File: windows/src/engine/kmcomapi/processes/keyboard/Keyman.System.Process.KPUninstallKeyboardLanguage.pas [hunk 2]
```diff
@@ -0,0 +1,382 @@
...
+    procedure UninstallTip(const KeyboardID: string; LangID: Word); overload;
+
+    /// <summary>Uninstalls all TIPs for the current user for a given keyboard</summary>
+    /// <param name="KeyboardID">The ID of the keyboard</param>
+    procedure UninstallTip(const KeyboardID: string); overload;
+
+  protected
+    /// <summary>Unregisters a TIP from the Local Machine registry</summary>
+    /// <param name="BCP47Tag">This should be a BCP47 tag</param>
+    /// <param name="KeyboardID">The ID of the keyboard</param>
```
Comment (darcywong00) [inline windows/src/engine/kmcomapi/processes/keyboard/Keyman.System.Process.KPUninstallKeyboardLanguage.pas:39]: Does it matter in the Delphi IDE if the parameters are listed in reverse order?

File: windows/src/engine/kmcomapi/com/packages/keymanpackagefile.pas [hunk 3]
```diff
@@ -167,22 +167,31 @@ function TKeymanPackageFile.Get_ID: WideString;
...
 procedure TKeymanPackageFile.Install(Force: WordBool);
+var
+  o: TKPInstallPackageOptions;
 begin
   with TKPInstallPackage.Create(Context) do
   try
-    Execute(FFileName, Force, True);
+    o := [ipLegacyRegisterAndInstallProfiles];
+    if Force then
+      Include(o, ipForce);
```
Comment (darcywong00) [inline windows/src/engine/kmcomapi/com/packages/keymanpackagefile.pas:177]: Where does `ipForce` come from?

File: windows/src/engine/kmcomapi/keymanapi_TLB.pas [hunk 4]
```diff
@@ -1592,7 +1604,7 @@ interface
 // *********************************************************************//
   IKeymanKeyboardFile2Disp = dispinterface
     ['{EDE4326B-51F4-42D5-8251-B20B71993EC8}']
-    function Install2(Force: WordBool; InstallDefaultLanguage: WordBool): IKeymanKeyboardInstalled; dispid 403;
+    function Install2(Force: WordBool): IKeymanKeyboardInstalled; dispid 1610809344;
```
Comment (darcywong00) [inline windows/src/engine/kmcomapi/keymanapi_TLB.pas:1607]: wow, that's quite a diff from `403`

File: windows/src/engine/kmcomapi/processes/keyboard/Keyman.System.Process.KPUninstallKeyboardLanguage.pas (hunk 2 above)
Comment (mcdurdin) [reply windows/src/engine/kmcomapi/processes/keyboard/Keyman.System.Process.KPUninstallKeyboardLanguage.pas:39]: It won't matter but I should fix it for clarity!

File: windows/src/engine/kmcomapi/com/packages/keymanpackagefile.pas (hunk 3 above)
Comment (mcdurdin) [reply windows/src/engine/kmcomapi/com/packages/keymanpackagefile.pas:177]: `ipForce` is defined in the enumeration in kpinstallpackage.pas:

```
TKPInstallPackageOptions = set of (ipForce, ipLegacyRegisterAndInstallProfiles);
```

File: windows/src/engine/kmcomapi/keymanapi_TLB.pas (hunk 4 above)
Comment (mcdurdin) [reply windows/src/engine/kmcomapi/keymanapi_TLB.pas:1607]: The dispids are automatically generated; perhaps I should tidy them up...

**Review by darcywong00 — APPROVED** [review]
lgtm

**Pull request merged**